src/app/clusters/thread-border-router-management-server/thread-border-router-management-server.cpp

```diff
@@ -0,0 +1,351 @@
+    return CHIP_NO_ERROR;
+}
+
+Optional<uint64_t> ServerInstance::ReadActiveDatasetTimestamp()
+{
+    uint64_t activeDatasetTimestampValue = 0;
+    Thread::OperationalDataset activeDataset;
+    if ((mDelegate->GetDataset(activeDataset, Delegate::DatasetType::kActive) == CHIP_NO_ERROR) &&
+        (activeDataset.GetActiveTimestamp(activeDatasetTimestampValue) == CHIP_NO_ERROR))
+    {
+        return MakeOptional(activeDatasetTimestampValue);
+    }
+    return NullOptional;
+}
+
+CHIP_ERROR ServerInstance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    if (aPath.mClusterId != ThreadBorderRouterManagement::Id)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    VerifyOrDie(mDelegate);
+    CHIP_ERROR status = CHIP_NO_ERROR;
+    switch (aPath.mAttributeId)
+    {
+    case Globals::Attributes::FeatureMap::Id: {
+        BitFlags<Feature> featureMap;
+        ReadFeatureMap(featureMap);
+        status = aEncoder.Encode(featureMap);
+        break;
+    }
+    case Attributes::BorderRouterName::Id: {
+        char borderRouterNameBuf[kBorderRouterNameMaxLength] = { 0 };
+        MutableCharSpan borderRouterName(borderRouterNameBuf);
+        status = ReadBorderRouterName(borderRouterName);
+        // If there are any internal errors, the status will be returned and the client will get an error report.
+        if (status == CHIP_NO_ERROR)
+        {
+            status = aEncoder.Encode(borderRouterName);
+        }
+        break;
+    }
+    case Attributes::BorderAgentID::Id: {
+        uint8_t borderAgentIDBuf[kBorderAgentIdLength] = { 0 };
+        MutableByteSpan borderAgentID(borderAgentIDBuf);
+        status = ReadBorderAgentID(borderAgentID);
+        if (status == CHIP_NO_ERROR)
+        {
+            status = aEncoder.Encode(borderAgentID);
+        }
+        break;
+    }
+    case Attributes::ThreadVersion::Id: {
+        uint16_t threadVersion = mDelegate->GetThreadVersion();
+        status                 = aEncoder.Encode(threadVersion);
+        break;
+    }
+    case Attributes::InterfaceEnabled::Id: {
+        bool interfaceEnabled = mDelegate->GetInterfaceEnabled();
+        status                = aEncoder.Encode(interfaceEnabled);
+        break;
+    }
+    case Attributes::ActiveDatasetTimestamp::Id: {
+        Optional<uint64_t> activeDatasetTimestamp = ReadActiveDatasetTimestamp();
+        status = activeDatasetTimestamp.HasValue() ? aEncoder.Encode(DataModel::MakeNullable(activeDatasetTimestamp.Value()))
+                                                   : aEncoder.EncodeNull();
+        break;
+    }
+    default:
+        break;
+    }
+    return status;
+}
+
+void ServerInstance::CommitSavedBreadcrumb()
+{
+    if (mBreadcrumb.HasValue())
+    {
+        GeneralCommissioning::SetBreadcrumb(mBreadcrumb.Value());
+    }
+    mBreadcrumb.ClearValue();
+}
+
+void ServerInstance::OnActivateDatasetComplete(uint32_t sequenceNum, CHIP_ERROR error)
```

Comment:
On success/failure, this should be used to mark the attributes dirty (if necessary) so that reporting occurs.

Also, it is important that this call only occur from Matter context, so if this is called from within the openthread task, this could lead to significant race issues.

Reply:
Done